metal/resource_metal_connection.go

```diff
@@ -0,0 +1,195 @@
+				Type:          schema.TypeString,
+				Optional:      true,
+				Computed:      true,
+				Description:   "Metro where to the connection will be created",
+				ConflictsWith: []string{"facility"},
+				ForceNew:      true,
+			},
+			"redundancy": {
+				// TODO: remove ForceNew and do Update, https://github.com/packethost/packngo/issues/270
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Connection redundancy - redundant or primary",
+				ForceNew:    true,
+			},
+			"type": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Connection type - dedicated or shared",
+				ForceNew:    true,
+			},
+			"organization_id": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "ID of the organization responsible for the connection",
+				ForceNew:    true,
+			},
+			"project_id": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "ID of the project where the connection is scoped to, only used for type == \"shared\"",
+				ForceNew:    true,
+			},
+			"description": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Description of the connection resource",
+				// TODO: remove ForceNew and do Update, https://github.com/packethost/packngo/issues/270
+				ForceNew: true,
+			},
+			"status": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Status of the connection resource",
+			},
+			"token": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Fabric Token from the [Equinix Fabric Portal](https://ecxfabric.equinix.com/dashboard)",
+			},
+			"speed": {
+				Type:        schema.TypeInt,
+				Computed:    true,
+				Description: "Port speed in bits per second",
+			},
+			"ports": {
+				Type:        schema.TypeList,
+				Elem:        connectionPortSchema(),
+				Computed:    true,
+				Description: "List of connection ports - primary (`ports[0]`) and secondary (`ports[1]`)",
+			},
+		},
+	}
+}
+
+func resourceMetalConnectionCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*packngo.Client)
+
+	facility, facOk := d.GetOk("facility")
+	metro, metOk := d.GetOk("metro")
+
+	if !(metOk || facOk) {
+		return fmt.Errorf("You must set either metro or facility")
+	}
+
+	project, projectOk := d.GetOk("project_id")
+	connType := packngo.ConnectionType(d.Get("type").(string))
+
+	if connType == packngo.ConnectionShared && !projectOk {
+		return fmt.Errorf("When you create a \"shared\" connection, you must set project_id")
+	}
+	if connType == packngo.ConnectionDedicated && projectOk {
+		return fmt.Errorf("When you create a \"dedicated\" connection, you mustn't set project_id")
+	}
+
+	connReq := packngo.ConnectionCreateRequest{
+		Name:       d.Get("name").(string),
+		Redundancy: packngo.ConnectionRedundancy(d.Get("redundancy").(string)),
+		Type:       connType,
+	}
+
+	if connType == packngo.ConnectionShared {
```

Comment:
nitpic: Since Project, Metro, and Facility are all omitempty in ConnectionCreateRequest, we could just set them.

https://github.com/packethost/packngo/blob/0d47a079fa31f25dd263b50ec9d3d820f0ae10f0/connections.go#Lv77-L87 (speed, type, and redundancy should perhaps not have omitempty).